<a href="https://colab.research.google.com/github/catarina1532/python/blob/main/kiva_loans.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###Introduction
Kiva.org is an online crowdfunding platform to extend financial services to poor and financially excluded people around the world.

kiva_loans is the dataset that contains most of the informations about the loans (id of borrower, amount of loan, time of repayment, reason for borrowing...).

###Retrieving the Data

In [96]:
import kagglehub

#Download latest version
path = kagglehub.dataset_download("kiva/data-science-for-good-kiva-crowdfunding")

print("Path to dataset files:", path)

Using Colab cache for faster access to the 'data-science-for-good-kiva-crowdfunding' dataset.
Path to dataset files: /kaggle/input/data-science-for-good-kiva-crowdfunding


In [97]:
import os
#List all files and directories in the specified path
for item in os.listdir(path):
  print(item)

kiva_mpi_region_locations.csv
loan_themes_by_region.csv
loan_theme_ids.csv
kiva_loans.csv


In [98]:
import pandas as pd

#Replace 'path/to/your/file.csv' with the actual path to your file in Google Drive
loans = pd.read_csv(path+'/'+'kiva_loans.csv')

###Glimpse of Data

In [99]:
#Display the first 5 rows of the dataframe with .head()
display(loans.head())

id  funded_amount  loan_amount             activity          sector  \
0  653051          300.0        300.0  Fruits & Vegetables            Food   
1  653053          575.0        575.0             Rickshaw  Transportation   
2  653068          150.0        150.0       Transportation  Transportation   
3  653063          200.0        200.0           Embroidery            Arts   
4  653084          400.0        400.0           Milk Sales            Food   

                                                 use country_code   country  \
0            To buy seasonal, fresh fruits to sell.            PK  Pakistan   
1  to repair and maintain the auto rickshaw used ...           PK  Pakistan   
2  To repair their old cycle-van and buy another ...           IN     India   
3  to purchase an embroidery machine and a variet...           PK  Pakistan   
4                           to purchase one buffalo.           PK  Pakistan   

         region currency  partner_id                posted_time  \
0        Lahore      PKR       247.0  2014-01-01 06:12:39+00:00   
1        Lahore      PKR       247.0  2014-01-01 06:51:08+00:00   
2     Maynaguri      INR       334.0  2014-01-01 09:58:07+00:00   
3        Lahore      PKR       247.0  2014-01-01 08:03:11+00:00   
4  Abdul Hakeem      PKR       245.0  2014-01-01 11:53:19+00:00   

              disbursed_time                funded_time  term_in_months  \
0  2013-12-17 08:00:00+00:00  2014-01-02 10:06:32+00:00            12.0   
1  2013-12-17 08:00:00+00:00  2014-01-02 09:17:23+00:00            11.0   
2  2013-12-17 08:00:00+00:00  2014-01-01 16:01:36+00:00            43.0   
3  2013-12-24 08:00:00+00:00  2014-01-01 13:00:00+00:00            11.0   
4  2013-12-17 08:00:00+00:00  2014-01-01 19:18:51+00:00            14.0   

   lender_count                          tags borrower_genders  \
0            12                           NaN           female   
1            14                           NaN   female, female   
2             6  user_favorite, user_favorite           female   
3             8                           NaN           female   
4            16                           NaN           female   

  repayment_interval        date  
0          irregular  2014-01-01  
1          irregular  2014-01-01  
2             bullet  2014-01-01  
3          irregular  2014-01-01  
4            monthly  2014-01-01

In [100]:
#Print a summary of the dataframe with .info()
loans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 671205 entries, 0 to 671204
Data columns (total 20 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   id                  671205 non-null  int64  
 1   funded_amount       671205 non-null  float64
 2   loan_amount         671205 non-null  float64
 3   activity            671205 non-null  object 
 4   sector              671205 non-null  object 
 5   use                 666973 non-null  object 
 6   country_code        671197 non-null  object 
 7   country             671205 non-null  object 
 8   region              614405 non-null  object 
 9   currency            671205 non-null  object 
 10  partner_id          657698 non-null  float64
 11  posted_time         671205 non-null  object 
 12  disbursed_time      668809 non-null  object 
 13  funded_time         622874 non-null  object 
 14  term_in_months      671205 non-null  float64
 15  lender_count        671205 non-nul

In [101]:
#Description of the dataframe with .describe()
loans.describe()

id  funded_amount    loan_amount     partner_id  \
count  6.712050e+05  671205.000000  671205.000000  657698.000000   
mean   9.932486e+05     785.995061     842.397107     178.199616   
std    1.966113e+05    1130.398941    1198.660073      94.247581   
min    6.530470e+05       0.000000      25.000000       9.000000   
25%    8.230720e+05     250.000000     275.000000     126.000000   
50%    9.927800e+05     450.000000     500.000000     145.000000   
75%    1.163653e+06     900.000000    1000.000000     204.000000   
max    1.340339e+06  100000.000000  100000.000000     536.000000   

       term_in_months   lender_count  
count   671205.000000  671205.000000  
mean        13.739022      20.590922  
std          8.598919      28.459551  
min          1.000000       0.000000  
25%          8.000000       7.000000  
50%         13.000000      13.000000  
75%         14.000000      24.000000  
max        158.000000    2986.000000

###Data Preparation

####Checking for missing data

In [102]:
#Checking missing data in the dataframe
total = loans.isnull().sum().sort_values(ascending = False)
percent = (loans.isnull().sum()/loans.isnull().count()).sort_values(ascending = False)
missing_loans  = pd.concat([total, percent], axis=1, keys=['Total', 'Percent'])
missing_loans

Total   Percent
tags                171416  0.255385
region               56800  0.084624
funded_time          48331  0.072006
partner_id           13507  0.020124
use                   4232  0.006305
borrower_genders      4221  0.006289
disbursed_time        2396  0.003570
country_code             8  0.000012
activity                 0  0.000000
loan_amount              0  0.000000
funded_amount            0  0.000000
id                       0  0.000000
posted_time              0  0.000000
currency                 0  0.000000
sector                   0  0.000000
country                  0  0.000000
lender_count             0  0.000000
term_in_months           0  0.000000
repayment_interval       0  0.000000
date                     0  0.000000

####Currency Conversion to USD

In [103]:
#Identify unique currencies in the dataframe
unique_currencies = loans['currency'].unique()
print("Unique currencies:", unique_currencies)

Unique currencies: ['PKR' 'INR' 'KES' 'NIO' 'USD' 'TZS' 'PHP' 'PEN' 'XOF' 'LRD' 'VND' 'HNL'
 'MNT' 'COP' 'GTQ' 'TJS' 'BOB' 'YER' 'KHR' 'GHS' 'SLL' 'HTG' 'CLP' 'JOD'
 'UGX' 'BIF' 'IDR' 'GEL' 'UAH' 'EUR' 'ALL' 'CRC' 'XAF' 'TRY' 'AZN' 'DOP'
 'BRL' 'MXN' 'KGS' 'AMD' 'PYG' 'LBP' 'WST' 'ILS' 'RWF' 'ZMW' 'NPR' 'MZN'
 'ZAR' 'BZD' 'SRD' 'NGN' 'VUV' 'XCD' 'MWK' 'LAK' 'MMK' 'ZWD' 'MDL' 'SSP'
 'SBD' 'CNY' 'EGP' 'MGA' 'NAD' 'LSL' 'THB']


In [104]:
#Define exchange rates (illustrative values)
exchange_rates_to_usd = {
    'USD': 1.0, #United States Dollar
    'PKR': 0.0036, #Pakistani Rupee
    'INR': 0.012, #Indian Rupee
    'KES': 0.0076, #Kenyan Shilling
    'NIO': 0.027, #Nicaraguan Córdoba
    'TZS': 0.00038, #Tanzanian Shilling
    'PHP': 0.017, #Philippine Peso
    'PEN': 0.27, #Peruvian Sol
    'XOF': 0.0016, #West African CFA franc
    'LRD': 0.0067, #Liberian Dollar
    'VND': 0.000039, #Vietnamese Dong
    'HNL': 0.041, #Honduran Lempira
    'MNT': 0.00029, #Mongolian Tögrög
    'COP': 0.00026, #Colombian Peso
    'GTQ': 0.13, #Guatemalan Quetzal
    'TJS': 0.11, #Somoni Tajique
    'BOB': 0.14, #Bolivian Boliviano
    'YER': 0.0045, #Yemeni Rial
    'KHR': 0.000031, #Cambodian Riel
    'GHS': 0.075, #Ghanaian Cedi
    'SLL': 0.000049, #Sierra Leonean Leone
    'HTG': 0.0075, #Haitian Gourde
    'CLP': 0.0011, #Chilean Peso
    'JOD': 1.41, #Jordanian Dinar
    'UGX': 0.00026, #Ugandan Shilling
    'BIF': 0.00011, #Burundian Franc
    'IDR': 0.000063, #Indonesian Rupiah
    'GEL': 0.36, #Georgian Lari
    'UAH': 0.033, #Ukrainian Hryvnia
    'EUR': 1.07, #Euro
    'ALL': 0.011, #Albanian Lek
    'CRC': 0.0019, #Costa Rican Colón
    'XAF': 0.0016, #Central African CFA franc
    'TRY': 0.031, #Turkish Lira
    'AZN': 0.59, #Azerbaijani Manat
    'DOP': 0.017, #Dominican Peso
    'BRL': 0.20, #Brazilian Real
    'MXN': 0.045, #Mexican Peso
    'KGS': 0.011, #Kyrgyzstani Som
    'AMD': 0.0026, #Armenian Dram
    'PYG': 0.00013, #Paraguayan Guarani
    'LBP': 0.000011, #Lebanese Pound
    'WST': 0.23, #Samoan Tala
    'ILS': 0.29, #Israeli New Sheqel
    'RWF': 0.00077, #Rwandan Franc
    'ZMW': 0.054, #Zambian Kwacha
    'NPR': 0.0089, #Nepalese Rupee
    'MZN': 0.016, #Mozambican Metical
    'ZAR': 0.068, #South African Rand
    'BZD': 0.015, #Belizean Dollar
    'SRD': 0.028, #Surinamese Dollar
    'NGN': 0.00072, #Nigerian Naira
    'VUV': 0.0083, #Vanuatu Vatu
    'XCD': 0.023, #East Caribbean Dollar
    'MWK': 0.00057, #Malawian Kwacha
    'LAK': 0.000047, #Lao Kip
    'MMK': 0.00048, #Myanmar Kyat
    'ZWD': 0.0031, #Zimbabwean Dollar
    'MDL': 0.00062, #Moldovan Leu
    'SSP': 0.0077, #South Sudanese Pound
    'SBD': 0.12, #Solomon Islands Dollar
    'CNY': 0.14, #Chinese Yuan Renminbi
    'EGP': 0.021, #Egyptian Pound
    'MGA': 0.00022, #Malagasy Ariary
    'NAD': 0.015, #Namibian Dollar
    'LSL': 0.00015, #Lesotho Loti
    'THB': 0.027, #Thai Baht
}

#Create new columns for USD converted amounts
loans['loan_amount_usd'] = loans.apply(
    lambda row: row['loan_amount'] * exchange_rates_to_usd.get(row['currency'], 0),
    axis=1)

loans['funded_amount_usd'] = loans.apply(
    lambda row: row['funded_amount'] * exchange_rates_to_usd.get(row['currency'], 0),
    axis=1)

#Display the first 5 rows with the new USD columns
display(loans[['currency', 'loan_amount', 'funded_amount', 'loan_amount_usd', 'funded_amount_usd']].head())

currency  loan_amount  funded_amount  loan_amount_usd  funded_amount_usd
0      PKR        300.0          300.0             1.08               1.08
1      PKR        575.0          575.0             2.07               2.07
2      INR        150.0          150.0             1.80               1.80
3      PKR        200.0          200.0             0.72               0.72
4      PKR        400.0          400.0             1.44               1.44

###Data Exploration

In [105]:
#Select the distinct countries from the dataframe using the .unique() method on the 'country' column
distinct_countries = loans['country'].unique()
print("Distinct Countries:", distinct_countries)

Distinct Countries: ['Pakistan' 'India' 'Kenya' 'Nicaragua' 'El Salvador' 'Tanzania'
 'Philippines' 'Peru' 'Senegal' 'Cambodia' 'Liberia' 'Vietnam' 'Iraq'
 'Honduras' 'Palestine' 'Mongolia' 'United States' 'Mali' 'Colombia'
 'Tajikistan' 'Guatemala' 'Ecuador' 'Bolivia' 'Yemen' 'Ghana'
 'Sierra Leone' 'Haiti' 'Chile' 'Jordan' 'Uganda' 'Burundi' 'Burkina Faso'
 'Timor-Leste' 'Indonesia' 'Georgia' 'Ukraine' 'Kosovo' 'Albania'
 'The Democratic Republic of the Congo' 'Costa Rica' 'Somalia' 'Zimbabwe'
 'Cameroon' 'Turkey' 'Azerbaijan' 'Dominican Republic' 'Brazil' 'Mexico'
 'Kyrgyzstan' 'Armenia' 'Paraguay' 'Lebanon' 'Samoa' 'Israel' 'Rwanda'
 'Zambia' 'Nepal' 'Congo' 'Mozambique' 'South Africa' 'Togo' 'Benin'
 'Belize' 'Suriname' 'Thailand' 'Nigeria' 'Mauritania' 'Vanuatu' 'Panama'
 'Virgin Islands' 'Saint Vincent and the Grenadines'
 "Lao People's Democratic Republic" 'Malawi' 'Myanmar (Burma)' 'Moldova'
 'South Sudan' 'Solomon Islands' 'China' 'Egypt' 'Guam' 'Afghanistan'
 'Madagascar' 'N

In [106]:
#Select loans from specific countries using the .isin() string method
selected_countries = ['Pakistan', 'India']
loans_by_country = loans[loans['country'].isin(selected_countries)]
display(loans_by_country.head())

id  funded_amount  loan_amount             activity          sector  \
0  653051          300.0        300.0  Fruits & Vegetables            Food   
1  653053          575.0        575.0             Rickshaw  Transportation   
2  653068          150.0        150.0       Transportation  Transportation   
3  653063          200.0        200.0           Embroidery            Arts   
4  653084          400.0        400.0           Milk Sales            Food   

                                                 use country_code   country  \
0            To buy seasonal, fresh fruits to sell.            PK  Pakistan   
1  to repair and maintain the auto rickshaw used ...           PK  Pakistan   
2  To repair their old cycle-van and buy another ...           IN     India   
3  to purchase an embroidery machine and a variet...           PK  Pakistan   
4                           to purchase one buffalo.           PK  Pakistan   

         region currency  ...             disbursed_time  \
0        Lahore      PKR  ...  2013-12-17 08:00:00+00:00   
1        Lahore      PKR  ...  2013-12-17 08:00:00+00:00   
2     Maynaguri      INR  ...  2013-12-17 08:00:00+00:00   
3        Lahore      PKR  ...  2013-12-24 08:00:00+00:00   
4  Abdul Hakeem      PKR  ...  2013-12-17 08:00:00+00:00   

                 funded_time term_in_months lender_count  \
0  2014-01-02 10:06:32+00:00           12.0           12   
1  2014-01-02 09:17:23+00:00           11.0           14   
2  2014-01-01 16:01:36+00:00           43.0            6   
3  2014-01-01 13:00:00+00:00           11.0            8   
4  2014-01-01 19:18:51+00:00           14.0           16   

                           tags  borrower_genders repayment_interval  \
0                           NaN            female          irregular   
1                           NaN    female, female          irregular   
2  user_favorite, user_favorite            female             bullet   
3                           NaN            female          irregular   
4                           NaN            female            monthly   

         date loan_amount_usd funded_amount_usd  
0  2014-01-01            1.08              1.08  
1  2014-01-01            2.07              2.07  
2  2014-01-01            1.80              1.80  
3  2014-01-01            0.72              0.72  
4  2014-01-01            1.44              1.44  

[5 rows x 22 columns]

In [107]:
#Calculate the minimum and maximum loan amounts in USD
min_loan_amount_usd = loans['loan_amount_usd'].min()
max_loan_amount_usd = loans['loan_amount_usd'].max()

#Select rows where loan_amount_usd equals the minimum loan amount in USD
min_loan_countries_usd = loans[loans['loan_amount_usd'] == min_loan_amount_usd]['country'].unique()

#Select rows where loan_amount_usd equals the maximum loan amount in USD
max_loan_countries_usd = loans[loans['loan_amount_usd'] == max_loan_amount_usd]['country'].unique()

print(f"Minimum Loan Amount USD: {min_loan_amount_usd} is associated with countries: {', '.join(min_loan_countries_usd)}")
print(f"Maximum Loan Amount USD: {max_loan_amount_usd} is associated with countries: {', '.join(max_loan_countries_usd)}")

Minimum Loan Amount USD: 0.000275 is associated with countries: Lebanon
Maximum Loan Amount USD: 100000.0 is associated with countries: Haiti


Lebanon shows the smallest loan amount recorded (USD 0.000275). In contrast, Haiti has the largest loan amount (USD 100000).

In [108]:
loans[(loans.loan_amount_usd - loans.funded_amount_usd) < 0]

id  funded_amount  loan_amount       activity       sector  \
277188  924149          425.0        400.0  General Store       Retail   
338159  985199         3400.0       3000.0  Farm Supplies  Agriculture   

                                                      use country_code  \
277188  to buy beverages, rice, laundry detergent, sug...           MZ   
338159  to pay for wires for the grape orchard, cover ...           AM   

           country                             region currency  ...  \
277188  Mozambique                      Boane, Maputo      MZN  ...   
338159     Armenia  Hoktember village, Armavir region      USD  ...   

                   disbursed_time funded_time term_in_months lender_count  \
277188  2015-07-22 07:00:00+00:00         NaN           17.0           11   
338159  2015-10-29 07:00:00+00:00         NaN           38.0           84   

                                                     tags  borrower_genders  \
277188                             #Parent, user_favorite              male   
338159  #Elderly, #Job Creator, #Trees, #Vegan, user_f...              male   

       repayment_interval        date loan_amount_usd funded_amount_usd  
277188            monthly  2015-07-28             6.4               6.8  
338159            monthly  2015-11-25          3000.0            3400.0  

[2 rows x 22 columns]

There are 2 instances when funded amount is greater than loan amount.

**Which country has received the highest number of loans?**

In [109]:
#Calculate the number of loans per country and sort in descending order
loans_by_country_count = loans['country'].value_counts()

#Display the top 5 countries by loan count
display(loans_by_country_count.head())

country
Philippines    160441
Kenya           75825
El Salvador     39875
Cambodia        34836
Pakistan        26857
Name: count, dtype: int64

Conclusion: Philippines has more number of loans given, followed by Kenya and El Salvador.

**Which sector has the highest number of loans?**

In [110]:
#Calculate the number of loans per sector and sort in descending order
loans_by_sector_count = loans['sector'].value_counts()

#Display the top 5 sectors by loan count
display(loans_by_sector_count.head())

sector
Agriculture     180302
Food            136657
Retail          124494
Services         45140
Personal Use     36385
Name: count, dtype: int64

Conclusion: Sector wise, Agriculture has the highest number of loans, followed by Food and Retail, showing Kiva’s focus on essential economic activities.

**Which loan activity received the highest number of funded loans?**

In [111]:
#Calculate the number of loans per activity and sort in descending order
loans_by_activity_count = loans['activity'].value_counts()

#Display the top 5 activities by loan count
display(loans_by_activity_count.head())

activity
Farming                      72955
General Store                64729
Personal Housing Expenses    32448
Food Production/Sales        28106
Agriculture                  27023
Name: count, dtype: int64

Conclusion: Farming has the highest number of funded loans, followed by General Store.

**What is the most common use of loans among borrowers?**



In [112]:
#Calculate the number of loans per use and sort in descending order
loans_by_use_count = loans['use'].value_counts()

#Display the top 5 uses by loan count
display(loans_by_use_count.head())

use
to buy a water filter to provide safe drinking water for their family.    5217
to buy a water filter to provide safe drinking water for her family.      4082
To buy a water filter to provide safe drinking water for their family.    2141
to build a sanitary toilet for her family.                                1708
to build a sanitary toilet for her family                                 1599
Name: count, dtype: int64

Conclusion: Most popular use of loan is to buy a water filter to provide safe drinking water for their family.

**Which country received the most loan funding?**

In [113]:
countries_funded_amount_usd = loans.groupby('country')['funded_amount_usd'].mean().sort_values(ascending = False)
display(countries_funded_amount_usd.head())

country
Cote D'Ivoire    50000.000000
Mauritania       15000.000000
Bhutan            7812.500000
Afghanistan       7000.000000
Puerto Rico       4409.191176
Name: funded_amount_usd, dtype: float64

Conclusion: Cote D'Ivoire is more loan funded, followed by Mauritania, showing that countries with higher costs of living or developing financial sectors often receive larger loan amounts.

**Which sector received the highest loan amount?**

In [114]:
sectors_loan_amount_usd = round(loans.groupby(['sector'])['loan_amount_usd'].mean().sort_values(ascending=False))
display(sectors_loan_amount_usd.head())

sector
Entertainment    1177.0
Wholesale         625.0
Education         500.0
Services          464.0
Construction      435.0
Name: loan_amount_usd, dtype: float64

Conclusion: Entertainment sector is taking more loan, followed by Wholesale.

**Borrower Gender: Female VS Male**

In [115]:
loans['female_borrower_count']= loans.borrower_genders.apply(lambda x: str(x).split(', ').count('female'))
loans['male_borrower_count']= loans.borrower_genders.apply(lambda x: str(x).split(', ').count('male'))

In [116]:
borrowers_percent = dict()
borrowers_percent['female'] = 100*loans['female_borrower_count'].sum()/(loans['female_borrower_count'].sum() + loans['male_borrower_count'].sum())
borrowers_percent['male'] = 100*loans['male_borrower_count'].sum()/(loans['female_borrower_count'].sum() + loans['male_borrower_count'].sum())
print('Female borrowers percentage: {:.2f} %'.format(borrowers_percent['female']))
print('Male borrowers percentage: {:.2f} %'.format(borrowers_percent['male']))

Female borrowers percentage: 79.58 %
Male borrowers percentage: 20.42 %


Conclusion: Approximately 80% borrower are Female and approximately 20% borrowers are Male.